robot/packages/utils.go

```diff
@@ -271,7 +272,8 @@ func commonCleanup(logger logging.Logger, expectedPackageDirectories map[string]
 			}
 			_, expectedToExist := expectedPackageDirectories[packageDirName]
 			_, expectedStatusFileToExist := expectedPackageDirectories[strings.TrimSuffix(packageDirName, statusFileExt)]
-			if !expectedToExist && !expectedStatusFileToExist {
+			_, expectedFirstRunSuccessFileToExist := expectedPackageDirectories[strings.TrimSuffix(packageDirName, config.FirstRunSuccessExt)]
+			if !expectedToExist && !expectedStatusFileToExist && !expectedFirstRunSuccessFileToExist {
```

Comment:
this logic is necessary to prevent the cleanup routine from deleting the setup phase success marker file

Reply:
sounds good, will add docs. as for consequences - deleting the marker file results in first run potentially running again the first time a new version of the package is downloaded.

Reply:
I think the consequence is more interesting? In what circumstance would we clean up this directory, but then redownload into it?

Reply:
hm, i guess i don’t know all the circumstances when this happens. 

my understanding is that this directory is cleaned when a module/version is no longer in the config but otherwise everything persists. @abe-winter is this understanding correct or am i missing some subtleties.

Reply:
> In what circumstance would we clean up this directory, but then redownload into it?

dgottlieb are you saying you know of this case? or just thinking defensively

maximpertsov speaking from memory, this cleanup logic deletes inner folders but not the directory itself (unless there's something at the bottom that deletes the whole dir when empty)

I am not now nor ever have I been particularly clear on the overall file layout of the ~/.viam dir, even though I have messed with it a lot -- I get confused btwn the data, download, and unpack dirs.

don't let me be a merge blocker, but if you want to sync + document the ~/.viam subfolders at some point, would not mind taking an hour

Reply:
> dgottlieb are you saying you know of this case? or just thinking defensively

Not thinking defensively. The code written here is already being defensive. But it's not communicating the reason why.

I was just hoping (perhaps wishfully) that we could write it down. Such if this defensive code turned out to be problematic for one scenario, we'd have more confidence about the case this was meant to guard against and preserve that behavior.

I mean, I suspect* this code is referring to a scenario of:
* User configures some (module, version) to run
* The package is downloaded successfully and the module runs for some time
* The user configures away the (module, version). The package is deleted except:
  * for the meta.json file and
  * the "first run" file.

It's not clear to me what bad thing would happen if the meta.json file were to be deleted. But it does seem right to keep the "first run" file.

But that's just based on @maximpertsov 's original (pr) comment:
> this logic is necessary to prevent the cleanup routine from deleting the setup phase success marker file

It's not even obvious to me that's even what the code is doing. I don't see the code preserving certain files. I just see it deleting an entire directory.

Reply:
(side note - thank you for all of the context - I'm going to address this documentation in a separate PR - mainly because I don't know exactly what to document here yet)

Reply:
my understanding is that the variables are not named well/haven't been renamed to suit the newer types of files we have added. Looking at a .viam dir with a tarball module right now, the module is unpacked into .viam/packages/data/module. That should correspond to a `packageTypeDir`. 

We read everything in the data/module dir with `packageDirs, err := os.ReadDir(packageTypeDirName)`. each packageDir in packageDirs can either be a directory, status.json, or .first_run_success file. In my case, there is a mod v1 directory (the extracted tarball), mod v1 status.json, and mod v1 first_run_succeeded file.

Looping through every packageDir entry, we decide whether or not to delete it based on if it is still being managed by the package manager. So if the package manager is supposed to be managing mod v2, all those previously listed files would be deleted based on the checks in L273-276. The first check checks if the file/dir name matches a module being managed, and the other two checks if the file/dir name matches a module being managed after the suffix has been removed.

The behavior now makes sense to me (if we start with v1, first_run succeeds and then we go to v2 and then back to v1, we should probably run v1's first_run again).

packageDir should probably have a different name haha